packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts

```diff
@@ -0,0 +1,188 @@
+    events: ICertificateEvent[];
+    cert: never;
+}
+interface FromReadModel extends ConstructorParams {
+    cert: ICertificateReadModel;
+    events: never;
+}
+export class CertificateAggregate {
+    private certificate: ICertificateReadModel | null = null;
+
+    private constructor(params: FromEvents | FromReadModel) {
+        if (params.events) {
+            params.events
+                .sort((first, second) => {
+                    return first.internalCertificateId - second.internalCertificateId;
+                })
+                .forEach((event) => this.apply(event));
+        } else {
+            this.certificate = params.cert;
+        }
+    }
+
+    public apply(event: ICertificateEvent): void {
+        switch (event.type) {
+            case CertificateEventType.Issued:
+                this.issue(event);
+                break;
+            case CertificateEventType.Transferred:
+                this.transfer(event as CertificateTransferredEvent);
+                break;
+            case CertificateEventType.Claimed:
+                this.claim(event as CertificateClaimedEvent);
+                break;
+        }
+    }
+
+    public getCertificate(): ICertificateReadModel | null {
+        return this.certificate;
+    }
+
+    public static fromReadModel(certificate: ICertificateReadModel): CertificateAggregate {
+        return new CertificateAggregate({ cert: certificate } as FromReadModel);
+    }
+
+    public static fromEvents(events: ICertificateEvent[]): CertificateAggregate {
+        return new CertificateAggregate({ events: events } as FromEvents);
+    }
+
+    private issue(event: ICertificateEvent): void {
+        if (this.isIssued()) {
+            throw new CertificateErrors.Issuance.CertificateAlreadyIssued(
+                event.internalCertificateId
+            );
+        }
+        const payload = event.payload as IIssueCommand<unknown>;
+
+        this.certificate = {
+            internalCertificateId: event.internalCertificateId,
+            blockchainCertificateId: null,
+            claims: [],
+            creationBlockHash: '',
+            creationTime: Math.floor(Date.now() / 1000),
+            deviceId: payload.deviceId,
+            generationStartTime: Math.floor(payload.fromTime),
+            generationEndTime: Math.floor(payload.toTime),
+            metadata: payload.metadata as any,
+            owners: {
+                [payload.toAddress]: payload.energyValue
+            },
+            claimers: {}
+        };
+    }
+
+    private transfer(event: CertificateTransferredEvent): void {
+        this.validateTransfer(event);
+
+        const { fromAddress, toAddress, energyValue } = event.payload;
+
+        const fromBalance = BigNumber.from(this.certificate!.owners[fromAddress] ?? '0');
+        const toBalance = BigNumber.from(this.certificate!.owners[toAddress] ?? '0');
+
+        const transferVolume = BigNumber.from(energyValue ?? '0');
```

Comment:
So please check this - I think if energyValue is not given you transfer all available energy, not zero. It would make no sense.